File: 2685-count-the-number-of-complete-components/2685-count-the-number-of-complete-components.py
class Solution:
    def countCompleteComponents(self, n: int, edges: List[List[int]]) -> int:
        adj_list = defaultdict(list)
        visited = [False] * n
        ans = 0

        for u, v in edges:
            adj_list[u].append(v)
            adj_list[v].append(u)

        def dfs(node):
            nonlocal v, e
            visited[node] = True
            v += 1
            e += len(adj_list[node])

            for neighbor in adj_list[node]:
                if not visited[neighbor]:
                    dfs(neighbor)

        for i in range(n):
            if not visited[i]:
                v = 0
                e = 0
                dfs(i)

                if e == v * (v - 1):
                    ans += 1

        return ans

File: 2685-count-the-number-of-complete-components/2685-count-the-number-of-complete-components.cpp
class Solution {
public:

    void dfs(int u,unordered_map<int,vector<int>> &adj, vector<bool> &visited, int &v, int &e){
        visited[u]=true;
        v++;
        e+=adj[u].size();
        for(int it : adj[u]){
            if (!visited[it]){
                dfs(it,adj,visited,v,e);
            }
        }
    }

    int countCompleteComponents(int n, vector<vector<int>>& edges) {
        vector<bool>visited(n,false);
        int ans=0;
        unordered_map<int,vector<int>>adj;
        for(auto &edge : edges){
            int u=edge[0];
            int v=edge[1];
            
            adj[u].push_back(v);
            adj[v].push_back(u);
        }

        for(int i=0;i<n;i++){
           if(!visited[i]) {

                int v = 0;
                int e = 0;

                dfs(i, adj, visited, v, e);

                if(e == v*(v-1))
                    ans++;
            }
        }
        return ans;
    }
};